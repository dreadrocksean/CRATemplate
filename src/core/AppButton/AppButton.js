import React from 'react';
import PropTypes from 'prop-types';

import styles from './appButton.module.scss';

const AppButton = ({ value, type, children, onClick }) => {
  return (
    <button onClick={onClick} type={type} className={styles.button}>
      {value || children}
    </button>
  );
};

AppButton.propTypes = {
  value: PropTypes.string.isRequired,
  type: PropTypes.string,
  children: PropTypes.string,
  onClick: PropTypes.func,
};

export default AppButton;
