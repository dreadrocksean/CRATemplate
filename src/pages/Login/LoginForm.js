import React, { Component } from "react";
import PropTypes from "prop-types";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";

import AppText from "../../core/AppText";
import TextInput from "../../core/formElements/TextInput";
import AppButton from "../../core/AppButton";

import Logo from "../../assets/images/SDlogo.png";

import styles from "./login.module.scss";

const appName = process.env.REACT_APP_APPNAME;

class LoginForm extends Component {
  constructor(props) {
    super(props);
    this.state = { username: "", password: "", submitText: "Log In" };
    this.interval = null;
  }

  componentWillUnmount() {
    this.cancelLoadingAnimation();
  }

  componentDidUpdate(prevProps) {
    if (this.props.loading !== prevProps.loading) {
      if (this.props.loading) this.beginLoadingAnimation();
      else
        this.cancelLoadingAnimation(() =>
          this.setState({ submitText: "Log In" })
        );
    }
  }

  cancelLoadingAnimation(cb = () => {}) {
    clearInterval(this.interval);
    this.interval = null;
    cb();
  }

  beginLoadingAnimation() {
    let dotCount = 0;
    if (this.interval) return;

    this.interval = setInterval(() => {
      const dots = [...Array(dotCount)].map(() => " .").join("");
      this.setState({ submitText: `Logging In${dots}` });
      dotCount = ++dotCount % 4;
    }, 400);
  }

  handleChange = (e, field) => {
    this.setState({ [field]: e.target.value });
  };

  handleSubmit = e => {
    e.preventDefault();
    this.props.handleSubmit({
      username: this.state.username,
      password: this.state.password
    });
  };

  render() {
    const { isAuthenticated, location, loading } = this.props;
    const { password } = this.state;
    return (
      <Card
        className={`${styles.loginFormWrapper} ${
          loading ? styles.loading : ""
        }`}
      >
        <CardContent>
          <div className={styles.formTitle}>
            {isAuthenticated ? (
              <div>
                <h2>Session Expired</h2>
                <p>
                  For your security, please re-enter your password to continue.
                </p>
              </div>
            ) : (
              <div>
                <a className={styles.logo} href={location}>
                  <img src={Logo} alt="MetroNet Logo" width="150" />
                </a>
                <AppText>
                  <h2>{appName}</h2>
                </AppText>
              </div>
            )}
          </div>

          <form
            className={styles.loginForm}
            data-validate-wrapper
            onSubmit={this.handleSubmit}
          >
            {!isAuthenticated ? (
              <TextInput
                type="username"
                value={this.state.username}
                handleChange={this.handleChange}
              />
            ) : (
              <div>
                <TextInput
                  type="username"
                  hidden
                  value={this.props.username} // From AuthStore
                  handleChange={this.handleChange}
                />
                <div>{this.props.username}</div>
              </div>
            )}
            <TextInput
              type="password"
              value={password}
              handleChange={this.handleChange}
            />

            <div className={styles.formActions}>
              {isAuthenticated ? (
                <AppButton type="submit" value="Continue" />
              ) : (
                <AppButton type="submit" value={this.state.submitText} />
              )}
            </div>
          </form>
        </CardContent>
      </Card>
    );
  }
}

LoginForm.propTypes = {
  location: PropTypes.string,
  isAuthenticated: PropTypes.bool,
  loading: PropTypes.bool,
  handleSubmit: PropTypes.func,
  username: PropTypes.string
};

export default LoginForm;
