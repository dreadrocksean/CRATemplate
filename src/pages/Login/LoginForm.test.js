import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import renderer from 'react-test-renderer';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import LoginForm from './LoginForm';
import AppText from '../../core/AppText';
import TextInput from '../../core/formElements/TextInput';
import AppButton from '../../core/AppButton';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<LoginForm />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('LoginForm component renders the todo correctly', () => {
  it('renders correctly', () => {
    const rendered = renderer.create(<LoginForm />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});

test('Expect component to be called at least once', () => {
  const wrapper = shallow(<LoginForm />);
  expect(wrapper.find(Card).length).toEqual(1);
  expect(wrapper.find(CardContent).length).toEqual(1);
  expect(wrapper.find(AppText).length).toEqual(1);
  expect(wrapper.find(TextInput).length).toEqual(2);
  expect(wrapper.find(AppButton).length).toEqual(1);
  expect(wrapper.find('.loginForm').length).toEqual(1);
});
