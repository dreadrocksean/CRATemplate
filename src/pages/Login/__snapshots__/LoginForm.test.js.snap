// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component renders the todo correctly renders correctly 1`] = `
<div
  className="MuiPaper-root-31 MuiPaper-elevation1-34 MuiPaper-rounded-32 MuiCard-root-30 loginFormWrapper "
>
  <div
    className="MuiCardContent-root-58"
  >
    <div
      className="formTitle"
    >
      <div>
        <a
          className="logo"
        >
          <img
            alt="MetroNet Logo"
            src="SDlogo.png"
            width="150"
          />
        </a>
        <div>
          <h2>
            Inventory Manager
          </h2>
        </div>
      </div>
    </div>
    <form
      className="loginForm"
      data-validate-wrapper={true}
      onSubmit={[Function]}
    >
      <div
        className="textInput"
      >
        <label>
          Username
          <input
            autoComplete="username"
            format="text"
            name="Username"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        className="textInput"
      >
        <label>
          Password
          <input
            autoComplete="current-password"
            format="password"
            name="Password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        className="formActions"
      >
        <button
          className="button"
          type="submit"
        >
          Log In
        </button>
      </div>
    </form>
  </div>
</div>
`;
